includes/Core/Permissions/Permissions.php

```diff
@@ -248,26 +299,112 @@ private function map_meta_capabilities( array $caps, $cap, $user_id, $args ) {
 			}
 
 			if ( ! in_array( $cap, array( self::AUTHENTICATE, self::SETUP ), true ) ) {
-				// For regular users, require being authenticated. TODO: Take $user_id into account.
-				$prevent_access = ! $this->authentication->is_authenticated();
-
+				// For regular users, require being authenticated.
+				if ( ! Feature_Flags::enabled( 'dashboardSharing' ) && ! $this->is_user_authenticated( $user_id ) ) {
+					return array_merge( $caps, array( 'do_not_allow' ) );
+				}
 				// For admin users, also require being verified. TODO: Take $user_id into account.
-				if ( ! $prevent_access && user_can( $user_id, self::SETUP ) ) {
-					$prevent_access = ! $this->authentication->verification()->has();
+				if ( user_can( $user_id, self::SETUP ) && ! $this->authentication->verification()->has() ) {
+					return array_merge( $caps, array( 'do_not_allow' ) );
 				}
 
 				// For all users, require setup to have been completed.
-				if ( ! $prevent_access ) {
-					$prevent_access = ! $this->authentication->is_setup_completed();
+				if ( ! $this->authentication->is_setup_completed() ) {
+					return array_merge( $caps, array( 'do_not_allow' ) );
 				}
+			}
+		}
+
+		if ( in_array( $cap, self::get_dashboard_sharing_capabilities(), true ) ) {
+			$caps = array_merge( $caps, $this->check_dashboard_sharing_capability( $cap, $user_id, $args ) );
+		}
+
+		return $caps;
+	}
+
+	/**
+	 * Checks a dashboard sharing capability based on rules of dashboard sharing.
+	 *
+	 * @param string $cap     Capability to be checked.
+	 * @param int    $user_id User ID of the user the capability is checked for.
+	 * @param array  $args    Additional arguments passed to check a meta capability.
+	 * @return array Array with a 'do_not_allow' element if checks fail, empty array if checks pass.
+	 */
+	private function check_dashboard_sharing_capability( $cap, $user_id, $args ) {
+		// TODO remove this check when Dashboard Sharing feature flag is removed.
+		if ( ! Feature_Flags::enabled( 'dashboardSharing' ) ) {
+			return array( 'do_not_allow' );
+		}
+
+		$sharing_settings_object = new Module_Sharing_Settings( new Options( $this->context ) );
```

Comment:
That is what I had originally named this variable 😄  But then changed it since I wanted to differentiate it from the `$sharing_settings` primitive array a few lines below which holds the array of settings! In many PHP projects in the past, we used to capitalise object variable names or use the class name to differentiate them from a simple primitive. Like it is easy to differentiate between `$Module_Sharing_Settings` to be an object and say `$module_sharing_settings` to be a simple array of settings. 
`module_sharing_settings` is definitely the standard way to name this in our codebase, so I guess I will change it but wanted to know your thoughts if the `sharing_settings` declared below wouldn't sound confusing?

Reply:
Thanks, I do think that the current combo of `$module_sharing_settings` / `$sharing_settings`, while less than ideal, is probably the best we can reasonably do in this scenario.